### DISCUSSION TITLE
What is the desired behavior of existing alias

### DISCUSSION THREAD

**Discussion by B1Z0N** [tree-sitter/tree-sitter#2567] [Q&A] [answered]
If I create an alias like `alias('HelloWorld', $.myrule)` and `myrule` is my actual rule somewhere in `rules` array, then what is the **desired** behavior for this? 

I mean, when I am looking at `src/node-types.json` it shows me:

```json
      "types": [
        {
          "type": "myrule",
          "named": true
        },
      ]
```

Even though this alias `myrule` and that declared `myrule` is different. 

Request for comments: how this should behave? Perhaps we should bring some warning/error when this thing happens?

**Comment by amaanq** [comment, marked as answer]
that's intended, querying for myrule should match either case (or both) if the query is valid for either case.

**Comment by ahlinc** [comment]
![image](https://github.com/tree-sitter/tree-sitter/assets/14666676/3e6dc9bd-c15f-4f8a-8503-cb21b31c8200)

It behaves as a named rule but due to aliasing with totally different content and yes aliasing needs to be used thoughtfully by grammar authors but I think there is no case to prohibit or warn about something.

**Reply by quentin** [reply]
Hi,
If I understand this answer correctly, using `alias(X,Y)` it is possible to construct an AST that is not well-formed with respect to `node-types.json` when the shape of `X` may not be a valid shape of `Y`. By shape I mean name and types of *fields* and *children*.

For instance `alias("HelloWorld", $.myRule)` is safe if `myRule` accepts `"HelloWorld"`. Following the example above, `myrule: $ => seq('a', 'b')` does not accept `"HelloWorld"` hence the resulting AST is not well-formed with respect to `node-types.json`.
